src/Builder/Encoder/OperatorEncoder.php

```diff
@@ -0,0 +1,161 @@
+<?php
+
+declare(strict_types=1);
+
+namespace MongoDB\Builder\Encoder;
+
+use LogicException;
+use MongoDB\Builder\Stage\GroupStage;
+use MongoDB\Builder\Type\Encode;
+use MongoDB\Builder\Type\OperatorInterface;
+use MongoDB\Builder\Type\Optional;
+use MongoDB\Codec\EncodeIfSupported;
+use MongoDB\Exception\UnsupportedValueException;
+use stdClass;
+
+use function array_key_exists;
+use function assert;
+use function get_object_vars;
+use function is_array;
+use function is_object;
+use function property_exists;
+use function sprintf;
+
+/** @template-extends AbstractExpressionEncoder<stdClass, OperatorInterface> */
+class OperatorEncoder extends AbstractExpressionEncoder
+{
+    /** @template-use EncodeIfSupported<stdClass, OperatorInterface> */
+    use EncodeIfSupported;
+
+    public function canEncode(mixed $value): bool
+    {
+        return $value instanceof OperatorInterface;
+    }
+
+    public function encode(mixed $value): stdClass
+    {
+        if (! $this->canEncode($value)) {
+            throw UnsupportedValueException::invalidEncodableValue($value);
+        }
+
+        switch ($value::ENCODE) {
```

Comment:
Related to an earlier question in `psalm-baseline.xml`, but should OperatorInterface have a method that returns the encoding type instead of utilizing a class constant?

Alternatively, if constants are being used for performance reasons, then can we define a constant with some invalid sentinel value on the base interface and require implementing classes to override it?

Reply:
A method would be an alternative, but for the time being I've added the constant to `OperatorInterface`. With a default value of `Encode::Single`, the effect of not overriding this constant would be failing tests for any operator that doesn't encode as single operator (which the majority of our operators do).